lib/internal/modules/esm/utils.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const {
+  ArrayIsArray,
+  SafeSet,
+  SafeWeakMap,
+  ObjectFreeze,
+} = primordials;
+
+const {
+  ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING,
+  ERR_INVALID_ARG_VALUE,
+} = require('internal/errors').codes;
+
+const { getOptionValue } = require('internal/options');
+
+const {
+  setImportModuleDynamicallyCallback,
+  setInitializeImportMetaObjectCallback
+} = internalBinding('module_wrap');
+const {
+  getModuleFromWrap,
+} = require('internal/vm/module');
+
+const callbackMap = new SafeWeakMap();
+function setCallbackForWrap(wrap, data) {
+  callbackMap.set(wrap, data);
+}
+
+let defaultConditions;
+function getDefaultConditions() {
+  if (defaultConditions === undefined) {
+    initializeDefaultConditions();
+  }
+  return defaultConditions;
+}
+
+let defaultConditionsSet;
+function getDefaultConditionsSet() {
+  if (defaultConditionsSet === undefined) {
+    initializeDefaultConditions();
+  }
+  return defaultConditionsSet;
+}
+
+function initializeDefaultConditions() {
+  const userConditions = getOptionValue('--conditions');
+  const noAddons = getOptionValue('--no-addons');
+  const addonConditions = noAddons ? [] : ['node-addons'];
+
+  defaultConditions = ObjectFreeze([
+    'node',
+    'import',
+    ...addonConditions,
+    ...userConditions,
+  ]);
```

Comment:
This is *still* run before any user code (or otherwise the array iterator deletion test would break) ;) Just because it's in a pre-loaded function doesn't mean it gets called lazily, we have plenty of time before loading any user code during pre-execution.

Reply:
Added some comments.